kernel/src/scan/log_replay.rs

```diff
@@ -220,32 +243,45 @@ impl AddRemoveDedupVisitor<'_> {
         // WARNING: It's not safe to partition-prune removes (just like it's not safe to data skip
         // removes), because they are needed to suppress earlier incompatible adds we might
         // encounter if the table's schema was replaced after the most recent checkpoint.
-        let partition_values = match &self.transform {
+        let (raw_partition_values, parsed_partition_values) = match &self.transform {
             Some(transform) if is_add => {
-                let partition_values =
+                let raw_partition_values =
                     getters[Self::ADD_PARTITION_VALUES_INDEX].get(i, "add.partitionValues")?;
-                let partition_values = self.parse_partition_values(transform, &partition_values)?;
-                if self.is_file_partition_pruned(&partition_values) {
+                let parsed_partition_values =
+                    self.parse_partition_values(transform, &raw_partition_values)?;
+                if self.is_file_partition_pruned(&parsed_partition_values) {
                     return Ok(false);
                 }
-                partition_values
+                (Some(raw_partition_values), parsed_partition_values)
             }
-            _ => Default::default(),
+            _ => (None, Default::default()),
         };
 
         // Check both adds and removes (skipping already-seen), but only transform and return adds
         if self.deduplicator.check_and_record_seen(file_key) || !is_add {
             return Ok(false);
         }
-        let transform = self
-            .transform
-            .as_ref()
-            .map(|transform| self.get_transform_expr(transform, partition_values))
-            .transpose()?;
-        if transform.is_some() {
-            // fill in any needed `None`s for previous rows
-            self.row_transform_exprs.resize_with(i, Default::default);
-            self.row_transform_exprs.push(transform);
+        if let Some(transform) = self.transform.as_ref() {
+            if let Some(raw_partition_values) = raw_partition_values {
+                // Convert to sorted Vec for cache key
+                let cache_key = self.partition_values_to_cache_key(&raw_partition_values);
+
+                let idx = if let Some(&cached_idx) = self.partition_cache.get(&cache_key) {
+                    // Cache hit - reuse existing transform
+                    cached_idx
+                } else {
+                    // Cache miss - build new transform
+                    let transform_expr =
+                        self.get_transform_expr(transform, parsed_partition_values)?;
+                    let idx = self.transform_cache.get_or_insert(transform_expr);
+                    self.partition_cache.insert(cache_key, idx);
+                    idx
+                };
```

Comment:
I think you could use [HashMap::entry](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry)?
```suggestion
                let idx = match self.partition_cache.entry(cache_key) {
                    Entry::Occupied(entry) => *entry.get(),
                    Entry::Vacant(entry) => {
                        // Cache miss - build new transform
                        let transform_expr =
                            self.get_transform_expr(transform, parsed_partition_values)?;
                        *entry.insert(self.transform_cache.get_or_insert(transform_expr))
                    }
                };
```